redisreceiver: add additional metrics

echlebek opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
This commit adds metrics to the redis receiver which are available to users that are using the telegraf collector. Providing greater parity here can help users that are attempting to switch from the telegraf collector to the otel collector.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #39859

<!--Describe what testing was performed and which tests were added.-->
#### Testing
TODO. Seeking assistance on what testing if any needs to be done here, as it's not clear to me if the generated tests are sufficient.

<!--Describe the documentation added.-->
#### Documentation
Documentation automatically added by mdatagen.

Commit:
redisreceiver: add additional metrics

This commit adds metrics to the redisreceiver that are present in the
telegraf collector. Improving parity between the telegraf collector and
the redisreceiver can help users that are moving between the otel
collector and the telegraf collector.

Fixes #39859

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
set disabled by default for new metrics

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

atoulme commented:
Please add a changelog. Please follow the guidance in the docs I gave you. Your PR lacks the code to actually capture those metrics.

Crucially, please check that none of the metrics you added are not already collected or cannot be computed from existing metrics by combining them.

echlebek commented:
@atoulme Will do, thank you.

atoulme commented:
Moving to draft while you address this feedback. Please mark ready for review once CI passes and the review comments are addressed.

Commit:
remove mem_fragmentation_bytes

This metrics is the delta of two other metrics.

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
Add changelog entry

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
Add metric_functions hooks

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
Run make generate

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
Add tracking_total_keys to test data

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
bump test expectation by 1 after adding to testdata

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
account for disabled datapoints in test

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
fix unit, run make generate

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
use input_type: string for cluster_enabled metric

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
Remove cluster_enabled metric

The cluster_enabled metric does not parse cleanly, so I have dropped it
for now.

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
fix testdata

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
add cluster_enabled metric back

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
improve changelog entry

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

echlebek commented:
I've hopefully covered all of the feedback given, thanks to all who review.

atoulme commented:
@hughesjj @dmitryax please review as codeowners.

echlebek commented:
Any feedback from code owners would be appreciated, thanks!

echlebek commented:
Any review here would be appreciated, even a rejection. Thanks!

echlebek commented on receiver/redisreceiver/metadata.yaml:
```diff
@@ -352,6 +352,34 @@ metrics:
+    enabled: false
+    description: "The sum in bytes of all overheads that the server allocated for managing its internal data structures"
+    unit: bytes
+    gauge:
+      value_type: int
```
I think that this observed value represents the total amount of memory overhead, so unless I'm mistaken a gauge is the correct representation?

echlebek commented:
Hi all, I'd like to re-open this PR and also request some discussion about it.

To re-introduce this PR's background: this PR is part of an effort in our organization to replace a library dependency on telegraf in our collector distribution. We're hoping to be able to remove that dependency without sacrificing too many features, and these metrics come into play as data that makes its way to many existing dashboards. We want to use only otel for all signals, but of course there are some spots that don't intersect nicely.

dmitryax commented on receiver/redisreceiver/metadata.yaml:
```diff
@@ -352,6 +352,34 @@ metrics:
+    gauge:
+      value_type: int
+
+  redis.stats.tracking_total_keys:
```
The naming is already not the same as what redis exposes, so it's not that easy to search anyways. All metrics emitted by the collector receivers have to align with the OTel naming guidelines: https://opentelemetry.io/docs/specs/semconv/general/naming. Please inspect other names as well and update them if needed

dmitryax commented:
@echlebek reopened. Please address comments and rebase the PR

Commit:
simplify tracking_total_keys metric name

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
Prefer non-monotonic sum to gauge

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
set stability level for cluster_enabled

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
run "make generate"

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

Commit:
fix changelog entry

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

atoulme commented on receiver/redisreceiver/dmesg.txt:
```diff
@@ -0,0 +1,4629 @@
+May 04 11:55:51 kernel: Linux version 6.13.11-200.fc41.x86_64 (mockbuild@fcf5690cacf1407ea15fadfe8861ad32) (gcc (GCC) 14.2.1 20250110 (Red Hat 14.2.1-7), GNU ld version 2.43.1-5.fc41) #1 SMP PREEMPT_DYNAMIC Thu Apr 10 19:02:09 UTC 2025
```
Is this file used in testing?

Commit:
remove accidentally included file

Signed-off-by: Eric Chlebek <echlebek@sumologic.com>

echlebek commented:
@dmitryax any further comments here?

JaredTan95 reviewed (approved):
LGTM
